server/api/roster/roster.controller.ts

```diff
@@ -172,28 +172,65 @@ class RosterController {
     });
 
     const columns = await Roster.getAllowedColumns(org, req.appRole!);
+    const errors: Array<{ error: string, edipi?: string, column?: RosterColumnInfo }> = [];
+    const existingEntries = await Roster.find({
+      where: `edipi IN (${roster.map(({edipi}) => `'${edipi}'`).join(',')})`
+    });
+
+    const onError = (error: Error, row?: RosterFileRow, column?: RosterColumnInfo) => {
+      errors.push({ column, edipi: row?.edipi, error: error.message });
+    }
+
     roster.forEach(row => {
-      if (!row.unit) {
-        throw new BadRequestError('Unable to add roster entries without a unit ID.');
-      }
-      const unit = orgUnits.find(u => row.unit === u.id);
-      if (!unit) {
-        throw new NotFoundError(`Unit with ID ${row.unit} could not be found in the group.`);
+      let unit: Unit | undefined;
+      try {
+        if (!row.unit) {
+          throw new BadRequestError('Unable to add roster entries without a unit ID.');
+        }
+        unit = orgUnits.find(u => row.unit === u.id);
+        if (!unit) {
+          throw new NotFoundError(`Unit with ID ${row.unit} could not be found in the group.`);
+        }
+        if (existingEntries.some(({edipi}) => edipi === row.edipi)) {
+          throw new BadRequestError(`Entry with EDIPI already exists.`);
+        }
+      } catch (error) {
+        onError(error, row);
       }
```

Comment:
In my initial pass, it did return. However, the lack of a return is intentional. The code catches all errors across all imported entities to avoid the "Upload -> Fix 1 issue -> Upload -> Fix 1 issue" cycle.